Make BEEFY aware of duplicated authorities

Klapeyron opened the pull request:
As a follow-up to https://github.com/paritytech/polkadot-sdk/issues/8705, we would like to propose an adjustment to the BEEFY RoundTracker, allowing it to count duplicated authorities as non-equal weight votes, in a similar way as handled by [GRANDPA's VoterSet](https://docs.rs/finality-grandpa/latest/finality_grandpa/voter_set/struct.VoterSet.html#method.new).

We choose this approach because adapting the interfaces related to OneSessionHandler would be significantly more invasive. Since GRANDPA already supports handling duplicated authorities in this manner, we decided to implement a similar solution here for consistency and simplicity.

This change should not affect how equivocations (double voting) are handled, nor does it alter the structure of commitments. It only impacts how votes from duplicated authorities are accounted for when calculating the threshold.

Klapeyron committed:
Make BEEFY aware of duplicated authorities

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

Klapeyron committed:
Additional docs for RoundTracker

Klapeyron committed:
Extended test

Klapeyron committed:
Moved weight calculation to Rounds

Klapeyron commented on substrate/client/consensus/beefy/src/round.rs:
```diff
@@ -78,6 +91,8 @@ pub enum VoteImportResult<B: Block, AuthorityId: AuthorityIdBound> {
 	Stale,
 }
 
+pub(crate) type VoteWeight = u32;
```
please help me with choosing right integer types for that

acatangiu replied:
u32 is ok

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

Klapeyron committed:
Moved missing weight handling outside

Klapeyron committed:
Replace key-value mapping with position based mapping

Klapeyron committed:
Revert "Replace key-value mapping with position based mapping"

This reverts commit 6b50a77a38961790eb06eabad66cf02f8b941c00.

Klapeyron committed:
Revert "Moved missing weight handling outside"

This reverts commit d60d8db69966aa9b41b47c42f79a174705472e87.

Klapeyron committed:
Persist only authorities with non 1 weight

Klapeyron commented:
I have adjusted the implementation to persist weights only for authorities that have voting weight different than 1, it should allow to store empty vector in case that a chain is not using that feature

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

Klapeyron committed:
Testcase scenario for non-equal weight rounds

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

acatangiu commented on substrate/client/consensus/beefy/src/round.rs:
```diff
@@ -126,6 +163,14 @@ where
 
+	/// Return voting weight associated with given authority or default 1 in case mapping does not exist
+	pub(crate) fn vote_weight(&self, authority: &AuthorityId) -> VoteWeight {
+		match self.voting_weights.binary_search_by(|(auth, _)| auth.cmp(authority)) {
```
binary search only works if the vec is sorted, which it isn't - you need to sort it when building it

Klapeyron replied:
during construction of `voting_weights` I am folding to `BTreeMap` which should handle sorting and deduplication for me:
```rust
			// We are sorting and deduplicating elements, which is later used by vote_weight function
			.fold(BTreeMap::new(), |mut acc, item| {
				*acc.entry(item.to_owned()).or_insert(0) += 1;
				acc
			})
```

acatangiu replied:
why not just keep `BTreeMap` instead of `Vec`? these assumptions are brittle, one might change that folding code without knowing there is a hidden assumption that the folding during build produces sorted collection.

Klapeyron replied:
oh I think it evaluated and now it is not necessary, changed to BTreeMap, thanks!

acatangiu commented:
cc @serban300

Klapeyron commented:
yesterday I accidentally opened `justification.rs` and I think that `verify_with_validator_set` should be probably adjusted as well:
```rust
if signatories.len() >= crate::round::threshold(validator_set.len())
```
@acatangiu what do you think?

Klapeyron committed:
Draft of aggregating votes during sync

Klapeyron committed:
Revert "Draft of aggregating votes during sync"

This reverts commit cace5acbcb2bac1ca0ff94efdef6e82b8e923b23.

Klapeyron committed:
Additional testcase

Klapeyron committed:
Replaced Vec with BTreeMap

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

serban300 reviewed (commented):
> yesterday I accidentally opened `justification.rs` and I think that `verify_with_validator_set` should be probably adjusted as well:
> 
> ```rust
> if signatories.len() >= crate::round::threshold(validator_set.len())
> ```

Yes, this code would definitely need to be adjusted as well

> Maybe at this point it is worth to make a breaking change and introduce a ValidatorSet with weight inside:
> 
> ```rust
> pub struct ValidatorSet<AuthorityId, VoteWeight> {
> 	/// Public keys of the validator set elements with associated vote weight
> 	validators: Vec<(AuthorityId, VoteWeight)>,
> 	/// Identifier of the validator set
> 	id: ValidatorSetId,
> }
> ```

Yes, that would be more natural. Adding the `voting_weights` seems more like a patch. It's just that it might lead to a lot of changes. I don't know what would be best.

Klapeyron committed:
Rename

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

Klapeyron committed:
Fixed formatting

Klapeyron committed:
Updated aux storage

Klapeyron committed:
Added prdoc

acatangiu commented:
@drskalman how does this look like in the context of BLS? can you please review this PR?

Klapeyron commented:
@serban300 @acatangiu may as ask for your review of this changes?

Klapeyron committed:
Renamed already_voted to accumulated_votes_weight

Klapeyron committed:
Removed vote_weight function and differentiation of weighted authorities

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

acatangiu commented on substrate/client/consensus/beefy/src/round.rs:
```diff
@@ -179,8 +202,9 @@ where
 		}
 
 		// add valid vote
+		let vote_weight = self.voting_weights.get(&vote.id).copied().unwrap_or(1);
```
if `self.voting_weights.get(&vote.id)` returns `None`, this function should return an invalid validator.id error rather than default to `vote_weight = 1`

Klapeyron committed:
Changed handling of missing weight

Klapeyron committed:
Storage migration from v4 to v5

serban300 commented:
No, I won't manage this week, sorry. Anyway I'll try in the following period. It's on my list

Klapeyron committed:
Merge branch 'master' of github.com:input-output-hk/polkadot-sdk into beefy_non_equal_weight_aware

Klapeyron committed:
Changed v4 -> v5 migration to use try_from

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

Klapeyron commented on substrate/client/consensus/beefy/src/aux_schema.rs:
```diff
@@ -18,18 +18,101 @@
+		pub(crate) votes: BTreeMap<AuthorityId, <AuthorityId as RuntimeAppPublic>::Signature>,
+		pub(crate) accumulated_votes_weight: VoteWeight,
+	}
+}
```
@serban300 quick reminder to recheck the PR when you can

serban300 replied:
Sorry, it's just a very busy period. Not sure when I'll be able to take another look on the PR. Will try next week, or the following one

justinfrevert replied:
Hi, @serban300 is there any chance you can take a look at this soon?

serban300 replied:
Yes, now it's a better time. I will try tomorrow

serban300 replied:
`try_from` looks good. But I still think that defining these migrated structures in the same file where the updated structures are (for example `PersistedStateV4` in the same file with `PersistedState`) would be easier to follow. Also they could access the private fields of the other structure, so we wouldn't need functions like `decode_from_migrated`

Klapeyron replied:
https://github.com/paritytech/polkadot-sdk/pull/8964/changes/8bbc1a0a1af872593c017c1544121b920c67bdb6

serban300 commented on substrate/client/consensus/beefy/src/round.rs:
```diff
@@ -34,30 +34,43 @@ use std::collections::BTreeMap;
+			None => {
+				return false;
+			},
+		}
```
Nit:
```suggestion
		self.accumulated_votes_weight = match self.accumulated_votes_weight.checked_add(vote_weight)
		{
			Some(sum) => sum,
			None => {
				return false;
			},
		};
```

Klapeyron replied:
https://github.com/paritytech/polkadot-sdk/pull/8964/commits/daf35aedad5b522187de4c6437af3f0bb29593e6

serban300 commented on substrate/client/consensus/beefy/src/round.rs:
```diff
@@ -161,6 +184,15 @@ where
 			return VoteImportResult::Invalid;
 		}
```
Do we still need this check ? If `vote.id` is not in the validators set, it will not be in the `voting_weights` either, right ?

Or it's safer to check this as well anyway ?

Klapeyron replied:
https://github.com/paritytech/polkadot-sdk/pull/8964/commits/7e37ade05f5c54fda34282f55fbdc9cdc2765ac5

serban300 commented on substrate/client/consensus/beefy/src/round.rs:
```diff
@@ -104,10 +120,17 @@ where
+			validator_set.validators().iter().fold(BTreeMap::new(), |mut acc, authority| {
+				*acc.entry(authority.to_owned()).or_insert(0) += 1;
+				acc
+			});
```
Nit: This logic is duplicated in `compute_voting_weights()` . Can we use the function here as well ?
Also could we define this logic on the ValidatorSet struct ?

Klapeyron replied:
https://github.com/paritytech/polkadot-sdk/pull/8964/commits/7e37ade05f5c54fda34282f55fbdc9cdc2765ac5

serban300 commented on substrate/client/consensus/beefy/src/aux_schema.rs:
```diff
@@ -104,16 +300,145 @@ pub(crate) mod tests {
+
+		// Between v4 and v5 we changed how vote weights are accumulated.
+		// Check this indirectly: with Alice(2) already present, a single Bob(1) vote should
+		// immediately conclude the round (threshold=3 for 3 validators).
```
I think it would be safest to check that all the old fields are the same and that the newly added `accumulated_votes_weight` is as expected as well.

Klapeyron replied:
https://github.com/paritytech/polkadot-sdk/pull/8964/commits/ee5f2af765377a8305410454b03746abe5559e5a

serban300 commented on substrate/client/consensus/beefy/src/aux_schema.rs:
```diff
@@ -73,7 +259,8 @@ where
 			warn!(target: LOG_TARGET,  "🥩 backend contains a BEEFY state of an obsolete version {v}. ignoring...")
 		},
-		Some(4) => {
+		Some(4) => return migrate_v4_to_v5::<B, _, AuthorityId>(backend),
```
Nit: If we want to also save the migrated state here, we should rename the function since `load_persistent` is not accurate anymore. And also we should mention this in the comment above. That it also migrates the persisted state to the latest one and saves it in memory.

Klapeyron replied:
https://github.com/paritytech/polkadot-sdk/pull/8964/commits/962222cecd8785da14fa9e803a5edbf7f3608bdf

serban300 commented on substrate/client/consensus/beefy/src/aux_schema.rs:
```diff
@@ -49,12 +203,44 @@ pub(crate) fn write_voter_state<B: BlockT, BE: AuxStore, AuthorityId: AuthorityI
+		&[],
+	)?;
+
+	load_decode::<_, PersistedState<B, AuthorityId>>(backend, WORKER_STATE_KEY)
```
```suggestion
    write_current_version(backend)?;
	write_voter_state(backend, &new_state)?;

	Ok(Some(new_state))
```

Klapeyron replied:
https://github.com/paritytech/polkadot-sdk/pull/8964/commits/75c731cf49f495be0c9bb527ba5dd2db34322f5f

Klapeyron replied:
I reverted it here https://github.com/paritytech/polkadot-sdk/pull/8964/commits/a2a6e06029db5a134deedab8c474bebab4565269, in the production code it makes commit not atomic, in test itself helper is not allowing to persist v4 version which we need there

Klapeyron replied:
I replaced it by helper doing both https://github.com/paritytech/polkadot-sdk/pull/8964/commits/82d2ef39235aff138bd0e85a8caf81224a7c5a8f

serban300 commented on substrate/client/consensus/beefy/src/aux_schema.rs:
```diff
@@ -49,12 +203,44 @@ pub(crate) fn write_voter_state<B: BlockT, BE: AuxStore, AuthorityId: AuthorityI
 	}
 }
 
+fn migrate_v4_to_v5<B, BE, AuthorityId>(
```
Nit: I don't think we need a separate function for this one. The logic is pretty simple, and seems to be called from only 1 place, so we could just move it directly in the function that uses it.

Klapeyron replied:
https://github.com/paritytech/polkadot-sdk/pull/8964/commits/dfceadb0256c921fb22d3695c0c88d5dcb607433

serban300 commented on substrate/client/consensus/beefy/src/aux_schema.rs:
```diff
@@ -18,18 +18,172 @@
+	})
+}
+
+fn accumulated_votes_weight<AuthorityId: AuthorityIdBound>(
```
Nit: I would define this method on `RoundTracker`. We could have a constructor `from_votes` for example.

Klapeyron replied:
https://github.com/paritytech/polkadot-sdk/pull/8964/commits/532463a11e1d5a01c9e00a859429124fd3b26511

serban300 commented on substrate/client/consensus/beefy/src/round.rs:
```diff
@@ -91,6 +106,7 @@ pub(crate) struct Rounds<B: Block, AuthorityId: AuthorityIdBound> {
 	>,
 	session_start: NumberFor<B>,
 	validator_set: ValidatorSet<AuthorityId>,
+	voting_weights: BTreeMap<AuthorityId, VoteWeight>,
```
`voting_weights` and `validator_set` are somewhat duplicate structures. Maybe we can't modify `ValidatorSet` to store a `BTreeMap` instead of a vec because it's probably also used in the runtime and other places. But could we have another `WeightedValidatorSet` (or some better name), does just that ?

Klapeyron replied:
https://github.com/paritytech/polkadot-sdk/pull/8964/commits/89758b0af26532db25af0ae538f25905098d9128

serban300 commented on substrate/client/consensus/beefy/src/aux_schema.rs:
```diff
@@ -18,18 +18,172 @@
+
+fn decode_from_migrated<T: Decode, U: Encode>(value: U) -> ClientResult<T> {
+	let encoded = value.encode();
+	T::decode_all(&mut &encoded[..])
```
encoding + re-decoding is not optimal. I would rather define a pub(crate) unchecked_from_parts() constructor for each struct that needs it.

Klapeyron replied:
https://github.com/paritytech/polkadot-sdk/pull/8964/changes/8bbc1a0a1af872593c017c1544121b920c67bdb6

serban300 reviewed (commented):
Just left a lot of nits mostly. The core idea looks good.

Klapeyron committed:
Unified compute_voting_weights

Klapeyron committed:
Update substrate/client/consensus/beefy/src/round.rs

Co-authored-by: Serban Iorga <serban300@gmail.com>

Klapeyron committed:
inlined migrate_v4_to_v5

Klapeyron committed:
full migration structure validation

Klapeyron committed:
renamed load_persistent to load_or_migrate_persistent

Klapeyron committed:
extracted from_votes for RoundTracker

Klapeyron committed:
extracted WeightedValidatorSet

Klapeyron committed:
fmt & unchecked_from_parts

Klapeyron committed:
fixup! Unified compute_voting_weights

Signed-off-by: Tomasz Bartos <tomasz.bartos@iohk.io>

Klapeyron committed:
fixup! extracted WeightedValidatorSet

Signed-off-by: Tomasz Bartos <tomasz.bartos@iohk.io>

Klapeyron committed:
moved PersistedStateV4 structs closer to referenced ones

Klapeyron committed:
Merge branch 'master' into beefy_non_equal_weight_aware

Klapeyron committed:
reused helper functions to persist state

Klapeyron committed:
Revert "reused helper functions to persist state"

This reverts commit 75c731cf49f495be0c9bb527ba5dd2db34322f5f.

Klapeyron committed:
fixup! reused helper functions to persist state

Signed-off-by: Tomasz Bartos <tomasz.bartos@iohk.io>

Klapeyron committed:
fixup! fixup! reused helper functions to persist state

Signed-off-by: Tomasz Bartos <tomasz.bartos@iohk.io>

Klapeyron committed:
fixup! fixup! fixup! reused helper functions to persist state

Signed-off-by: Tomasz Bartos <tomasz.bartos@iohk.io>

Klapeyron commented:
@serban300 I think I addressed all comments, may I ask you for next round there?

serban300 commented on substrate/client/consensus/beefy/src/aux_schema.rs:
```diff
@@ -49,14 +68,17 @@ pub(crate) fn write_voter_state<B: BlockT, BE: AuxStore, AuthorityId: AuthorityI
+///
+/// If the backend contains an older supported schema, migrate it to the latest schema and save the
+/// migrated state back to the aux-db.
+pub(crate) fn load_or_migrate_persistent<B, BE, AuthorityId: AuthorityIdBound>(
```
I think `load_and_migrate_persistent` would be more accurate

serban300 replied:
@Klapeyron could you also address this comment please ?

Klapeyron replied:
https://github.com/paritytech/polkadot-sdk/pull/8964/commits/c0cc22a0755ce22abe52d0ad10636415da216a1d

serban300 commented on substrate/client/consensus/beefy/src/round.rs:
```diff
@@ -78,6 +115,60 @@ pub enum VoteImportResult<B: Block, AuthorityId: AuthorityIdBound> {
+
+/// Compute per-authority voting weights from a validator set: each duplicate
+/// occurrence of the same authority contributes +1 to its weight.
+pub(crate) fn compute_voting_weights<AuthorityId: AuthorityIdBound>(
```
Now this doesn't seem to be needed anymore :D Seems to be called just in 1 place, so we could just move the logic there

serban300 commented on substrate/client/consensus/beefy/src/round.rs:
```diff
@@ -34,33 +35,69 @@ use std::collections::BTreeMap;
+			None => {
+				return false;
+			},
+		};
```
```suggestion
self.accumulated_votes_weight = self.accumulated_votes_weight.saturating_add(vote_weight);
```

Actually I think we should use `saturating_add` here. I don't think there is any reason to fail in case of overflow. If there is overflow, the accumulated weight will be > threashold, which is ok. But probably it will never happen.

serban300 commented on substrate/client/consensus/beefy/src/round.rs:
```diff
@@ -78,6 +115,60 @@ pub enum VoteImportResult<B: Block, AuthorityId: AuthorityIdBound> {
+	voting_weights: BTreeMap<AuthorityId, VoteWeight>,
+}
+
+impl<AuthorityId: AuthorityIdBound> WeightedValidatorSet<AuthorityId> {
```
Nit: Looks like all these methods defined here can be private

serban300 commented on substrate/client/consensus/beefy/src/round.rs:
```diff
@@ -34,33 +35,69 @@ use std::collections::BTreeMap;
+///
+/// Kept only to migrate v4 state to the current schema; see [`Rounds`]'s `TryFrom` impl.
+#[derive(Debug, Decode, Encode, PartialEq)]
+pub(crate) struct RoundTrackerV4<AuthorityId: AuthorityIdBound> {
```
Nit: Sorry, small thing, and not a strong preference, but I think the code would be easier to follow if we defined these structures in a different order. Mainly I would define VoteWeight above WeightedValidatorSet, where it's used. And WeightedValidatorSet above RoundTracker, where it's used. And RoundTrackerV4 above RoundTracker, since it's older. So something like this probably:

VoteWeight
WeightedValidatorSet
RoundTrackerV4
RoundTracker

serban300 reviewed (commented):
Looks good ! Just a few more nits

Klapeyron commented:
@serban300 today my access to forked repository via org expired, I am afraid that at the same time I lost also the write access to this branch. I am thinking how to address the comments, the simplest approach sounds like a quick follow-up PR, elsewhere I will have to close this one and setup from new fork, but rewriting of the commits may cause copyright issues. WDYT?

serban300 commented:
> @serban300 today my access to forked repository via org expired, I am afraid that at the same time I lost also the write access to this branch. I am thinking how to address the comments, the simplest approach sounds like a quick follow-up PR, elsewhere I will have to close this one and setup from new fork, but rewriting of the commits may cause copyright issues. WDYT?

Can you try to open another PR that includes only the addressed comments on top of this branch (so basically a PR that could be merged inside this one). And I might have enough permissions to merge it. I'm not sure. 

If not, then yes, the other option is to close this PR and open another one instead

Klapeyron commented:
> Can you try to open another PR that includes only the addressed comments on top of this branch (so basically a PR that could be merged inside this one). And I might have enough permissions to merge it. I'm not sure.

That will require opening a PR to a forked repository, so someone with write access there has to approve the merge

Klapeyron committed:
review changes

serban300 commented:
> > Can you try to open another PR that includes only the addressed comments on top of this branch (so basically a PR that could be merged inside this one). And I might have enough permissions to merge it. I'm not sure.
> 
> That will require opening a PR to a forked repository, so someone with write access there has to approve the merge



> > Can you try to open another PR that includes only the addressed comments on top of this branch (so basically a PR that could be merged inside this one). And I might have enough permissions to merge it. I'm not sure.
> 
> That will require opening a PR to a forked repository, so someone with write access there has to approve the merge

Ok, than opening a new PR is probably the only option. I don't know about the copyright issues, I'm not familiar with this aspect. Could you expand on this please ?

Klapeyron commented:
I managed to get a temporary access to this specific repository, I pushed https://github.com/paritytech/polkadot-sdk/pull/8964/changes/4b5dd105df43ac850106559639c01eeeb58f1287

Klapeyron committed:
fixup! review changes

Signed-off-by: Tomasz Bartos <tomasz.bartos@iohk.io>

Klapeyron committed:
fixup! review changes

Signed-off-by: Tomasz Bartos <tomasz.bartos@iohk.io>

serban300 commented on substrate/client/consensus/beefy/src/worker.rs:
```diff
@@ -284,6 +285,72 @@ pub(crate) struct PersistedState<B: Block, AuthorityId: AuthorityIdBound> {
+			ClientError::Backend("BEEFY DB is corrupted: invalid migrated voter oracle".into())
+		})
+	}
+}
```
One last nit: Please let's also move this above the definition of `VoterOracle`. And `PersistedStateV4` above `PersistedState`.

serban300 reviewed (commented):
Sorry, one more nit. Probably the last one

Klapeyron committed:
fixup! review changes

Signed-off-by: Tomasz Bartos <tomasz.bartos@iohk.io>

serban300 commented:
There seems to be a problem with the merge queue. I'll try again to merge the PR after that is fixed
